### ISSUE TITLE
Should annotations files with no annotations be saved?

### ISSUE THREAD

**Issue by indigoviolet** [bastibe/annotate.el#104]
I noticed the [source code](https://github.com/bastibe/annotate.el/blob/54ac759facadacbfea5c1e7c2975e2da6434cdda/annotate.el#L1317) appears to prevent this, but I see files containing no annotations like this

```
(("~/dev/retinanet-examples/csrc/cuda/utils.h" nil "5d6550d74a31d9b6987c158b8773dca3"))
```
Note that I have customized `annotate-file` to be local to the file. Is this a bug?

**Comment by cage2** [comment]
Hi!

Very good catch! Fixed in #102 (or so i hope!). ;-)

Bye!
C.

**Comment by indigoviolet** [comment]
@cage2 - Thank you for your work on this! I just pulled #102 and tested this, and now I see that the annotation file is still saved, but it contains the text `nil`.

**Comment by cage2** [comment]
Hi @indigoviolet !

I have checked the issue but the last version of the package seems not to be affected by this problem, may i ask to recheck using this use case?

1. set the  database annotation file targeting another file, or just make a backup of the db

``` lisp 
(setq annotate-file "/a/file")
```
2. annotate a file

3. save the annotated file

4.  check that the annotation database contains an entry like:

``` lisp
(("~/an/annotated/file" ((90530 90548 "annotation text" "annotated-text")) "hex-digits"))
```
5. delete the annotation from the annotated file
6.  save again the (no more) annotated file
7.  the annotation database file should contains just `nil`

Thank you!  
C.

**Comment by indigoviolet** [comment]
@cage2 That is the behavior I see as well. 

But I think it is a bit inconvenient even to have `nil` in the annotation database file, because it means that if I open a file with `annotate-mode` on, then saving that file will lead to an empty annotation database file even if I never added any annotations.

So there are two cases:

1. I never added any annotations and saved a buffer with `annotate-mode` on -> ideally no database file should be created
2. I removed all annotations from a buffer that previous had them, and the annotation database file was previously created -> in this case I don't feel that strongly but it would be nice to remove the ("empty" == `nil` containing) annotation database file.  

Anyway, I'm also fine with shipping #102 as is for now - you have improved a lot of things in that PR! This is not such a big deal for me.

**Comment by cage2** [comment]
Hi!

> @cage2 That is the behavior I see as well.

Sorry i did not read your previous message properly! :(

> But I  think it  is a  bit inconvenient  even to  have `nil`  in the
> annotation database  file, because it  means that  if I open  a file
> with `annotate-mode` on, then saving that file will lead to an empty
> annotation database file even if I never added any annotations.

>
> So there are two cases:
>

> 1.  I  never   added  any  annotations  and  saved   a  buffer  with
  `annotate-mode` on -> ideally no database file should be created

> 2. I removed  all annotations from a buffer that  previous had them,
  and the annotation  database file was previously created  -> in this
  case I don't feel  that strongly but it would be  nice to remove the
  ("empty" == `nil` containing) annotation database file.

Well honestly i would prefer to not overload meanings of the existence
of a  file as synonym for  "empty database", we could  rather use this
for: "this is  the first time annotate mode has  been activated". But,
anyway,  i think  my preference  is matter  of personal  tastes as  my
arguments are pretty weak as you can see :)

> Anyway, I'm also fine  with shipping #102 as is for  now - you have
> improved a lot of things in that PR! This is not such a big deal for
> me.

Thank you! I am going to merge the branch in a minute.
C.

**Comment by cage2** [comment]
fixed by #102

**Comment by indigoviolet** [comment]
Could we leave this open? I still think it's inconvenient for empty database files to be created when no annotations have been created - perhaps it's helpful to explain why:

- I have annotate-mode hooked to prog-mode.
- I have a per-file annotation database file by making `annotate-file` a buffer-local like `<filename>.annotate`

In this scenario, every single file I open and save creates an empty `.annotate` file, which I would like to avoid.

**Comment by cage2** [comment]
Hi!

Sure, i am reopening the issue! 
What you have written seems quite valid points to me. 👍

Bye!
C.

**Comment by cage2** [comment]
Hi @indigoviolet !

May i ask to check if PR #107 solve this issue?

Thank!
C.

**Comment by indigoviolet** [comment]
will try and check soon, thank you!

**Comment by cage2** [comment]
Hi!!

> will try and check soon, thank you!

Good! Looking forward for your comments!

Bye!
C.

**Comment by cage2** [comment]
Hi!

Fixed by #107 

Bye!
C.
